connaisseur/validate.py

```diff
@@ -72,41 +72,80 @@ def process_chain_of_trust(
     """
     Processes the whole chain of trust, provided by the notary server (`host`)
     for any given `image`. The 'root', 'snapshot', 'timestamp', 'targets' and
-    potentially 'targets/releases' are requested in this order and afterwards
-    validated, also according to the `policy_rule`.
+    potentially 'targets/releases' are requested and validated.
+    Additionally, it is checked whether all required delegations are valid.
 
     Returns the signed image targets, which contain the digests.
 
     Raises `NotFoundExceptions` should no required delegetions be present in
     the trust data, or no image targets be found.
     """
-    tuf_roles = ["root", "snapshot", "timestamp", "targets"]
     trust_data = {}
     key_store = KeyStore()
 
-    # get all trust data and collect keys (from root and targets), as well as
-    # hashes (from snapshot and timestamp)
+    tuf_roles = ["root", "snapshot", "timestamp", "targets"]
+
+    # Load all trust data
     for role in tuf_roles:
-        trust_data[role] = get_trust_data(host, image, TUFRole(role))
-        key_store.update(trust_data[role])
+        role_trust_data = get_trust_data(host, image, TUFRole(role))
+        trust_data[role] = role_trust_data
+
+    # Validate signature and expiry data of and load root file
+    # This does NOT conclude the validation of the root file. To prevent roleback/freeze attacks,
+    # the hash still needs to be validated against the snapshot file
```

Comment:
I am not following why the validation now has been expanded. That was not the case in the previous iteration of the commit, and it also does not seem to be the objective of the commit (which only is to change order). Am I missing something obvious?

Reply:
It hasn't really been expanded but the order has changed. Previously `validate` was called, which in turn called `validate_signature`, `validate_expiry` and `validate_hash`. But it did so by loading and "trusting" all DCT files and then using this data to validate said files. This commit now does validation in an order such that there's never (potentially) untrusted data is used to validate other files. But that necessitates that the order be changed, since e.g. the hash of th root.json can only be compared to snapshot state once the signature on snapshot has been validated, yet that requires first validating previous files (as good as it gets).

Reply:
However, yes it changes the original intent of the commit. It fixes an additional vector of attack where the security was based on targeted collision resistance instead of 2nd preimage resistance of the DCT hash function (still theoretical only)